File: src/main/java/com/pancisin/bookster/components/LicenseService.java
package com.pancisin.bookster.components;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.stream.Collectors;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

import com.pancisin.bookster.model.Conference;
import com.pancisin.bookster.model.Page;
import com.pancisin.bookster.model.UserSubscription;
import com.pancisin.bookster.model.enums.PageState;
import com.pancisin.bookster.model.enums.SubscriptionState;
import com.pancisin.bookster.repository.ConferenceRepository;
import com.pancisin.bookster.repository.PageRepository;
import com.pancisin.bookster.repository.UserSubscriptionRepository;

@Component
public class LicenseService {

	@Autowired
	private UserSubscriptionRepository usRepository;

	@Autowired
	private EmailService emailService;

	@Autowired
	private PageRepository pageRepository;

	@Autowired
	private ConferenceRepository conferenceRepository;

	@Scheduled(cron = "0 0 3 * * *")
//	 @Scheduled(fixedDelay = 10000)
//	@Transactional
	public void checkLicenses() {
		List<UserSubscription> newSubs = new ArrayList<UserSubscription>();

		List<UserSubscription> expired = usRepository.findExpirations();

		expired.stream().forEach(s -> {
			if (s.getState() == SubscriptionState.ACTIVE) {
				s.setState(SubscriptionState.EXPIRED);
				newSubs.add(createNew(s));
			} else if (s.getState() == SubscriptionState.NEW) {
				s.setState(SubscriptionState.UNPAID);

				Long user_id = s.getUser().getId();

				List<Page> pages = pageRepository.getByOwner(user_id).stream().map(p -> {
					p.setState(PageState.BLOCKED);
					return p;
				}).collect(Collectors.toList());

				pageRepository.save(pages);

				List<Conference> conferences = conferenceRepository.getByOwner(user_id).stream().map(c -> {
					c.setState(PageState.BLOCKED);
					return c;
				}).collect(Collectors.toList());

				conferenceRepository.save(conferences);
			}
		});

		usRepository.save(newSubs);
		newSubs.stream().forEach(us -> {
			emailService.sendInvoiceEmail(us);
		});

		usRepository.save(expired);
	}

	private UserSubscription createNew(UserSubscription previous) {
		UserSubscription subscription = new UserSubscription();
		subscription.setUser(previous.getUser());
		subscription.setSubscription(previous.getSubscription());
		subscription.setState(SubscriptionState.NEW);

		Calendar expiration = Calendar.getInstance();
		expiration.add(Calendar.MONTH, 1);
		subscription.setExpires(expiration);
		return subscription;
	}
}

File: src/main/kotlin/com/pancisin/bookster/services/LicenseService.kt
package com.pancisin.bookster.services

import com.pancisin.bookster.model.Notification
import com.pancisin.bookster.model.User
import java.util.ArrayList
import java.util.Calendar

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component

import com.pancisin.bookster.model.UserSubscription
import com.pancisin.bookster.model.enums.NotificationType
import com.pancisin.bookster.model.enums.PageState
import com.pancisin.bookster.model.enums.SubscriptionState
import com.pancisin.bookster.repository.ConferenceRepository
import com.pancisin.bookster.repository.PageRepository
import com.pancisin.bookster.repository.UserSubscriptionRepository

@Component
class LicenseService {

  @Autowired
  lateinit var usRepository: UserSubscriptionRepository

  @Autowired
  lateinit var emailService: EmailService

  @Autowired
  lateinit var pageRepository: PageRepository

  @Autowired
  lateinit var conferenceRepository: ConferenceRepository

  @Autowired
  lateinit var notificationService: NotificationService

  @Scheduled(cron = "0 0 3 * * *")
  fun checkLicenses() {
    val expired = usRepository.findExpirations()
    val newSubs = ArrayList<UserSubscription>()

    expired.forEach { s ->
      if (s.state === SubscriptionState.ACTIVE) {
        s.state = SubscriptionState.EXPIRED
        newSubs.add(createNew(s))
      } else if (s.state === SubscriptionState.NEW) {
        val user = s.user
        if (user != null) {
          s.state = SubscriptionState.UNPAID

          var pages = pageRepository.getByOwner(user.id)
          pages.addAll(conferenceRepository.getByOwner(user.id))

          pages = pages
            .filter { it.state === PageState.PUBLISHED || it.state === PageState.DEACTIVATED }
            .onEach { it.state = PageState.BLOCKED }

          pageRepository.save(pages)

          notificationService.notifyUser(user, Notification(
            recipient = user,
            code = "notification.license.unpaid",
            target = s.subscription?.name,
            type = NotificationType.DANGER
          ))
        }
      }
    }

    usRepository.save(newSubs)
    newSubs.stream().forEach { emailService.sendInvoiceEmail(it) }

    usRepository.save(expired)
  }

  private fun createNew(previous: UserSubscription) = UserSubscription(
    user = previous.user,
    subscription = previous.subscription,
    state = SubscriptionState.NEW,
    expires = Calendar.getInstance().apply {
      add(Calendar.MONTH, 1)
    }
  )
}
